### PR TITLE
improve attribute reliability for ground primitive materials

### PR DESCRIPTION
**Pull request by likangning93**
Fixes https://github.com/AnalyticalGraphicsInc/cesium/issues/6739.

Ground Primitive materials don't work the same across all devices I've tried and basically don't work at all on many mobile devices that claim support for all required extensions. This is most easily observed on iOS. There's more notes in the issue, but I'm pretty sure what's happening is that floating point textures are unreliable for large-magnitude values, which is bad because we pack encoded positions on the surface of the Earth to floats in the batch table.

The floats are all within very specific ranges, so this PR naively packs each to 4 uint8 values. It's *very* wasteful, but memory use is actually the same and the "unpack" op looks pretty cheap, so the only real negative here is the additional code and the salty, salty tears of any information theorists who might look in here.

Note that this doesn't fix depth-texture related bugs, but it should at least make substituting corridors for polylines on terrain possible on iOS devices. I'll have to open an issue to say that the feature is merely "buggy" instead of completely broken.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @likangning93!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/mtls-on-groundPrimitives-iOS/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

**Comment by likangning93** [comment]
snafu, update CHANGES.md in the wrong place...

**Comment by bagnell** [comment]
This change looks OK to me, but there is a large increase in the number of texture reads from the batch table. Can we detect the issue and fall back to this instead of using it as the default behavior?

**Comment by likangning93** [comment]
> Can we detect the issue and fall back to this instead of using it as the default behavior?

Sorry for the radio silence. Going to try adding a compute pass to check floating point texture precision, but I can't get to that for at least another few days.

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @likangning93! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @likangning93! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @likangning93! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @likangning93! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @likangning93! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @likangning93! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by likangning93** [comment]
@cesium-concierge stop

**Comment by hpinkos** [comment]
@likangning93 what's the plan for this PR?

**Comment by likangning93** [comment]
I'm planning on finishing this up in time for next release

**Comment by likangning93** [comment]
@lilleyse this is ready for review. State of the changes are now:
* app does a quick compute command on `Context` construction to assess if float texture precision is sufficient
* all code that works with high/low positions in the batch table for ground primitive materials can take either a path using values packed to uint8 or values in a conventional float texture

[Sandcastle for testing](http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/mtls-on-groundPrimitives-iOS/Build/Apps/Sandcastle/index.html#c=7VsLc9u4Ef4rbGY6kloFIilRpOz4Wsd2rp6JE4+sJu2cbzI0CUlo+CoI+nE3/u9dgA+RFCnTlKxkLuVMYpHEPr7FYrFYgLcmle586tgzTKlJPOlIOsEhiVxkUWwy/Dn3rvv7tSfBRfF/Ixyyz/CaXpjh1wOJ0Qj3Cy8/Ycrw/QefuqYTxg2uvcfe4bV37d2CzFuC7zAFaR6+SyV+Es+6HUvcnvgeA6GYdvpSIpjFelxS/5bYmB4UNM/Y3xHP9u+Q6YAKwAzY4HsmzR3fZBL/GVEsBRRbJCS+J4XRfE4sgj32N6kj/TVRDIUW9jBKiJEgnsW0V+T+0jFBx1gamUvdPyUAfqZ+5NmXlLiEkVuMwigIfMrCC24pAobo5rn3ehmuJfXv8paYRh4jLj6j1Ad7lNhKbspOMgGK5zMpEYRtCRCxJQmlwDHZHKyPOlzNx9TsAJuZ3sLBh/G9MOA09/Dam0eexbhlzCBwHo6dYGlemEEKoRukavSlGEUK4sr0bMsMmYORjS0HVOtWcugdStJgIL3zqbQki6UD/xjxFhL4Xo5FzDMThoAVNqnpWRil8IvOkymYqMOvuXlDiSUdSLln/FoZcO0Vv8xY40RKZRPRaw8BhredFGCnX90uAh+BngjrOfGLuOZCsENoIH6HgxjZl/f+wv9y4js+Rf8JFnVS+GUtTc/DXOUO7VQ3e1x/XHr0WBJg+W7gezA+agDYZD6PQqH6LbaGXQXJvSolhVV5q4J5kbgrKZvTKPn5mHlx0T3fRm6wlXeWGHDn/M69M7F3c/88TQl276E3YD3XDFDgLZ7hbxVa3IhuaIoo6bQXhbNhmIWMYm/BlkAsI+MbjLOSC6DkvkrnMMBWBO4udJWVihaemKM523wnoPhx64F5ssTWV0xvfJPa7UdnFZddzh+l0Ynm1Hdn4GTdTl5ypw4jj8lbgMuTA6o9gOIS69GA/4WE4W0QlViIcLqveLqKqZXj5+moIKJAISCAEmzJh5DyddPcu8Qc1juCHVtap+fDKqFvFgyzCaBW03ikBhvjW8Ng2Txg1gTNpzFuiInPmCJCP6IWyK5m0rF+c79kvsJvFphlXpJ/ee4FEZMKd+DYkPrXK15knvPIRM4pnpuRsxJXZH64mXe8MIld6JPpRBj4JssU9bSbc61+UWkUsh6iTzB3S1ME8HbJfVckajJS+zApjPl/aq8vpdkbv9fShzm9ek1lxZnekcSZSa9zYITHVMLoSX8BgXpTCcmcdZQOhORBU/JsUjyCuRwHYAllHSkPW1dpQxdWYHw1KjE/CBuJWBIPFq9hCDIMJGcMlyYNsLeKhRt4UQxm8zJr1YPrPG6Ypmui/anP2sf5HPFOE+b6mQsk1s1bp8VkqAWaIoPvYQnwdOq+OWQ3Xky2Te/OXBLy8slWa68KJv9ff+2gQrDVNIuzTmmKL9eNOwdo5dLwp5ZlFIMTsKdzmHtoovQ3t3kQ6yVtJ7nMt1j2pb3Iyco9irKXfM5VW8eApP7bdvQXyBsuf/YSdIViz4u3aUIhcvLnrR/izEUQJjrfRMSxL3w7cvA/qdPtHIchZuEgKT+HgzuuX1JZF1x7FWznog7vWQ/g6zJckNZVlOQ80JN3KOQ52K6tEJhu4BAW2bA2goyurbv8TMkWlYA89V7WylxgXcpxznu3PZYC+Q7BbOn3Qq1vlGfUes1UBHVsC+Web+oC+Y9h6ooB/NTMKGZEuWZK5FOh+vS01rxLP6QRr/0IWmPx7bv2R6zaZ3PX03CyHts5mliJ55btx3+csn3WC9vW7K9WWUz7oVnB5HtaUefU21/83bAv+4zeYZQEW2QdRfrdVubr06hYal0iNSXu+0Rue2AVTPazR5QTXAEwPj/zNmLM98JuEUrWiPmLhYOzAxifSEhuCCTbD91e3gmzYxsoXPp3vEBMI35aI31fPMaRNprDbJS2ekwXTyXRnwuUdfKr+ZeUoI3k5+xv2vbM950bGI3Yi7q/5ASKIzswXE581/VXNdAwGT2PfamyMd/eyo0/MDx2QCt4mcIu4hLU9V1wWGxZsemXYU6dtUSSQ5s5YuXu33oYqIS4lji+BMTi8qY9xNI6pyHENGUfTH3G/0p7wVxcZ7THXFpwrGP+NT429vyhcEl9C9sRhaCU7Kg8MRryO9gvPSiqdvu3GBuV2/4N/YdvFbww3PzmSXuUhV2UhuBEVWJ36HLKF0N8JexCBac97mIppyHwJI944Y4t5VjtMZaTrR3GgbdmiBtOiOJkolRKdV/CbmunRNtbbv24aEMP4ecc9oG1fOSwPdS1s4dNI9z6uvFFolxpY3WLSFfeYW0INN1Z2ke3Vu1otkdcubXZEHVcKtkH5vVKXHvEFSW5pqE9rXXsAXFViWOLKF9V69hhqL8godUw1MNaVMoWo+2MWL8UrEyYK5buxRzi+blz1UK+oROtbdq9GPDitueWkEuboDv0nbN7vm2H4/okP5PZeBGdtk847MWo64dPtzRsxVHUjca1fStyscfQArMzB/Ofbx/O7W6HxXbu9FDIHkDOHbHZUjqSOor8505N0Sf7VKdc98lCDT95Jr4wyqpPIe/Ubq66WfrqJ1/kXGDfxYw+nHsh4+UqXrLMU5Zelw/JpuRFqszWKXn396qdpBTAwRqdqIydmJTBc9M7ptR86FZX2X/Jic0Iht3X6lAzlKGBjJE+HMmqofSl10N9NJJ1FY20iTZS4K4vjaCVohlI05WRomtDXevVlP3r5ejqUEGjiQrU8lgexoI0RR4i3TDGY1lVDE1IknVZRRNd0/WxOh62kKRNFGAgq6qsjYYpInU0QbI60XVjpMWIVACMFGiiG6BFG0CyASzliaaBnuNYDkgfIpAmjwxDSU03GSvIGA9VgKlOer+uC6oSfiu+53vHJ3m26vszxyFB6BP7KqJz08LHqwrqp7Pp7OxfX959nF4cz0o1717+fGJO2KoAW3TNejFlHzVv/Fscjxweo3gBskaW5ZiQoc2JJQ5AzOLtgdS0a+/Q7Gw6PT7/kJboV4GjGKf+sOP6FC8ohnj2WtEnaDIB9zLiv+m9uP0hPEdUt5MmL+BQ/LNM4oFvmM4ln8H4tHK0PuhFr0xNG35Dr0DoGI5Ghj5UDAggMNxH/AA3BC8Ip7Kq6xDldD7+NYgJaDgZjSb6xFCVsZpOizmpHyn/GtZMBOeM+Q8M4rzFJWHWcuo7TsExFB7lJtpkPNYVFSKYZoCPqAoajoYTWdNkQ1dHY4VrBUqNJ5oyUQxjYowULVOh8OWt6WJqohAz/lVwvrNsHEKGKNQ7KBsqn6qsUBxUIMu1xJ49A5cI+ebfQW67h5L7ETo/PfswO5/9u7w3x+0lNAUT5RWH9zGAhePfYIQ988bBaVa72jO59moyhcP0TXHjCB6/6r96I1KQn2Jd/k5c/s2vFFGni9CAYZd/8yu2962vmCErFF8ovxmkRG9scisR++j6VekD6+tXsfPCm3nkOFfkN3z96qc3A2hfIHN80f8fYUQ75gNvslR+eh8/RAi9GcDtOlWSQOU4/g8)

File: Source/Renderer/Context.js [hunk 1]
```diff
@@ -586,6 +590,18 @@ define([
             }
         },
 
+        /**
+         * <code>true</code> returns if the context's floating point textures support 6 decimal places of precision.
```
Comment (lilleyse) [inline Source/Renderer/Context.js]:
```suggestion
         * Returns <code>true</code> if the context's floating point textures support 6 decimal places of precision.
```

File: Source/Renderer/checkFloatTexturePrecision.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    PixelDatatype,
+    Texture
+) {
+    'use strict';
+
+    /**
+     * Checks if the context's floating point textures support 6 decimal places of precision.
+     *
+     * @param {Context} context A context wrapping a gl implementation.
+     * @returns {Boolean} whether or not the context's floating point textures support 6 decimal places of precision
```
Comment (lilleyse) [inline Source/Renderer/checkFloatTexturePrecision.js]:
```suggestion
     * @returns {Boolean} Whether or not the context's floating point textures support 6 decimal places of precision
```

File: Source/Shaders/CheckFloatTexturePrecisionFS.glsl [hunk 3]
```diff
@@ -0,0 +1,7 @@
+uniform sampler2D u_floatTexture;
+
+void main() {
```
Comment (lilleyse) [inline Source/Shaders/CheckFloatTexturePrecisionFS.glsl]: Put curly brace on its own line for all shader code.

**Comment by likangning93** [comment]
@lilleyse updated!

**Comment by lilleyse** [comment]
Thanks!

**Comment by likangning93** [comment]
I'll update issues soon

**Comment by likangning93** [comment]
Updated title for https://github.com/AnalyticalGraphicsInc/cesium/issues/6735 since the remaining iOS artifacts in both ground polylines and ground primitive materials have the same underlying cause.

**Comment by likangning93** [comment]
The saga continues here: https://groups.google.com/forum/#!topic/webgl-dev-list/aFmFGeyRVRQ

**Comment by likangning93** [comment]
@lilleyse from the continuation of the saga over in the webgl dev google group, looks like `sampler2D` precision is not covered by the `precision highp float` hint, looks like that was all we needed.

I'll open a PR to roll this back and do the precision updates in the batch table in time for the next release.

**Comment by mramato** [comment]
@likangning93 are there any other areas of the code where we were assuming highp for sample2D that should be updated as well?

**Comment by likangning93** [comment]
Probably anywhere else we use floating point textures, I can take a look.

**Comment by mramato** [comment]
@likangning93 cool, maybe this will end up greatly improving compatibility across many devices :crossed_fingers: If it's something you aren't going to get to right away, please write up an issue so we don't forget about it.

**Pull request merged**